npm/generate_katas_content.js

```diff
@@ -64,19 +84,74 @@ function tryReadFile(filePath, errorPrefix) {
   try {
     content = readFileSync(filePath, "utf8");
   } catch (e) {
-    throw new Error(`${errorPrefix}: ${e}`);
+    throw new Error(`${errorPrefix}\n${e}`);
   }
   return content;
 }
 
-function getMissingProperties(properties, required) {
+function identifyMissingProperties(properties, required) {
   return required.filter((property) => !Object.hasOwn(properties, property));
 }
 
-function generateExampleSection(kataPath, properties) {
+function getSourceId(sourcePath, basePath) {
+  return relative(basePath, sourcePath).replaceAll(sep, "__");
+}
+
+function aggregateSources(paths, globalCodeSources) {
+  const codeSources = [];
+  for (const path of paths) {
+    const id = getSourceId(path, globalCodeSources.basePath);
+    if (!(id in globalCodeSources.sources)) {
+      const code = tryReadFile(path, "Could not read code dependency");
+      globalCodeSources.sources[id] = code;
+    }
+    codeSources.push(id);
+  }
+  return codeSources;
+}
+
+function parseMarkdown(markdown) {
+  const tokens = [];
+  const macroRegex = /@\[(?<type>\w+)\]\((?<json>\{.*?\})\)\n/gs;
+  let latestProcessedIndex = 0;
+  while (latestProcessedIndex < markdown.length) {
+    const match = macroRegex.exec(markdown);
+    if (match !== null) {
+      // If there is something between the last processed index and the start of the match that is not just whitespace,
+      // it represents a text token.
+      const delta = match.index - latestProcessedIndex;
+      if (delta > 0) {
+        const textToken = tryCreateMarkdownToken(
+          markdown.substring(latestProcessedIndex, match.index)
+        );
+        if (textToken !== null) {
```

Comment:
Should it ever be null at this point? If not, throw an error

Reply:
It can be null at this point if the text is just white space characters.